ivadomed/utils.py

```diff
@@ -38,27 +41,69 @@ def get_results(self):
 
 class Evaluation3DMetrics(object):
 
-    def __init__(self, fname_pred, fname_gt, params=None):
+    def __init__(self, fname_pred, fname_gt, params={}):
         self.fname_pred = fname_pred
         self.fname_gt = fname_gt
 
-        if not params is None:
-            pass
-
         self.data_pred = self.get_data(self.fname_pred)
         self.data_gt = self.get_data(self.fname_gt)
 
         self.px, self.py, self.pz = self.get_pixdim(self.fname_pred)
 
-        # Note: Some papers suggested to remove all lesion with less than 3 voxels: Todo?
+        self.bin_struct = generate_binary_structure(3, 2)  # 18-connectivity
+
+        # Remove small objects
+        if "removeSmall" in params:
+            size_min = params['removeSmall']['thr']
+            if params['removeSmall']['unit'] == 'vox':
+                self.size_min = size_min
+            elif params['removeSmall']['unit'] == 'mm3':
+                self.size_min = np.round(size_min / (self.px * self.py * self.pz))
+            else:
+                print('Please choose a different unit for removeSmall. Chocies: vox or mm3')
+                exit()
+
+            self.data_pred = self.remove_small_objects(data=self.data_pred)
+            self.data_gt = self.remove_small_objects(data=self.data_gt)
+        else:
+            self.size_min = 0
+
+        if "targetSize" in params:
+            self.size_rng_lst, self.size_suffix_lst = \
+                                    self._get_size_ranges(thr_lst=params["targetSize"]["thr"],
+                                                                unit=params["targetSize"]["unit"])
+
+            self.data_gt_per_size = self.label_per_size(self.data_gt)
+            label_gt_size_lst = list(set(self.data_gt_per_size[np.nonzero(self.data_gt_per_size)]))
+            self.data_pred_per_size = self.label_per_size(self.data_pred)
+            label_pred_size_lst = list(set(self.data_pred_per_size[np.nonzero(self.data_pred_per_size)]))
+            self.label_size_lst = [label_gt_size_lst + label_pred_size_lst,
+                                    ['gt']*len(label_gt_size_lst) + ['pred']*len(label_pred_size_lst)]
+
+        else:
+            self.label_gt_size_lst, self.label_gt_size_lst = []
```

Comment:
here it should probably be: self.label_gt_size_lst, self.label_size_lst = [], []